### PR TITLE
Dynamically determine GKE Version for Cluster and Node Pool Creation

### PR DESCRIPTION
**Pull request by Obliviour**
## Fixes / Features
- Determine the GKE version based on gcloud recommended versions
- Support addition node pools with the correct version (use the cluster's master version) instead of accidently using a new version by default
- Support --gke-version=x to override the default behavior.

This is super useful as xpk no longer has to change their gke version it uses whenever it disappears as a gke master version AND xpk will no longer by default recommend bad node pool versions.

## Testing / Documentation
Testing details.

- [ y ] Tests pass
- [ y ] Appropriate changes to documentation are included in the PR

### CODE REVIEW

File: xpk.py [hunk 1]
```diff
@@ -4072,6 +4076,183 @@ def install_nccl_on_cluster(args) -> int:
...
+  return_code, default_rapid_gke_version = run_command_for_value(
+      default_rapid_gke_version_cmd, f'{base_command_description} {task}.', args
+  )
+  if return_code != 0:
+    xpk_print(f'Unable to get server config for {task}.')
+    return return_code, None
+
+  task = ' valid master versions'
+  return_code, valid_master_versions = run_command_for_value(
+      valid_master_versions_cmd, f'{base_command_description} .', args
```
Comment (SurbhiJainUSC) [inline xpk.py:4128]: Did you miss {task}?

File: xpk.py [hunk 2]
```diff
@@ -4072,6 +4076,183 @@ def install_nccl_on_cluster(args) -> int:
...
+
+  task = 'default rapid gke version'
+  return_code, default_rapid_gke_version = run_command_for_value(
+      default_rapid_gke_version_cmd, f'{base_command_description} {task}.', args
+  )
+  if return_code != 0:
+    xpk_print(f'Unable to get server config for {task}.')
+    return return_code, None
+
+  task = ' valid master versions'
```
Comment (SurbhiJainUSC) [inline xpk.py:4126]: nit: unnecessary whitespace at the start.

File: xpk.py [hunk 3]
```diff
@@ -4072,6 +4076,183 @@ def install_nccl_on_cluster(args) -> int:
...
+  return_code, valid_master_versions = run_command_for_value(
+      valid_master_versions_cmd, f'{base_command_description} .', args
+  )
+  if return_code != 0:
+    xpk_print(f'Unable to get server config for {task}.')
+    return return_code, None
+
+  task = 'valid node versions'
+  return_code, valid_node_versions = run_command_for_value(
+      valid_node_versions_cmd, f'{base_command_description} .', args
```
Comment (SurbhiJainUSC) [inline xpk.py:4136]: nit: missed {task}

File: xpk.py [hunk 4]
```diff
@@ -4072,6 +4076,183 @@ def install_nccl_on_cluster(args) -> int:
...
+      " {print $2}' "
+  )
+  valid_node_versions_cmd = (
+      base_command
+      + " | awk '/validNodeVersions/,/END/' | awk '/-/,/-/ {print $2}' "
+  )
+  base_command_description = 'Determine server supported GKE versions for'
+
+  task = 'default rapid gke version'
+  return_code, default_rapid_gke_version = run_command_for_value(
```
Comment (SurbhiJainUSC) [inline xpk.py:4119]: Can we move these 3 run_command_for_value() into a single method that reduces human error in the commands, logs etc.
Something like this:
```
gke_version_command_map = {}
gke_version_command_map['default rapid gke version'] =  default_rapid_gke_version_cmd
for task_name, command in gke_version_command_map.items():
      return_code, gke_version = run_command_for_value(
           command, f'{base_command_description} {task_name}.', args)
      if return_code != 0:
         xpk_print(f'Unable to get server config for {task}.')
      return return_code, None
```

File: xpk.py [hunk 5]
```diff
@@ -4072,6 +4076,183 @@ def install_nccl_on_cluster(args) -> int:
...
+        f'\nRecommended GKE Version: {gke_server_config.valid_node_versions}'
+    )
+    xpk_print(
+        'Error: Planned GKE Version {master_gke_version} is not valid.'
+        f'Checks failed: Is Master Valid: {is_valid_master_version}'
+        f'\nIs Valid Node Version: {is_valid_node_version}'
+    )
+    xpk_print(
+        'Please select a gke version from the above list using --gke-version=x'
+        ' argument. or rely on the default gke version:'
```
Comment (SurbhiJainUSC) [inline xpk.py:4193]: nit: remove `.` after argument?

File: xpk.py [hunk 6]
```diff
@@ -4072,6 +4076,183 @@ def install_nccl_on_cluster(args) -> int:
...
+        ' {gke_server_config.valid_node_versions}'
```
Comment (SurbhiJainUSC) [inline xpk.py:4194]: Is this the default gke version? Isn't `valid_node_versions` a set of node pool gke version?

File: xpk.py [hunk 7]
```diff
@@ -4072,6 +4076,183 @@ def install_nccl_on_cluster(args) -> int:
...
+  # Override with user provide gke version if specified.
+  if args.gke_version is not None:
+    master_gke_version = args.gke_version
+  else:
+    master_gke_version = gke_server_config.default_rapid_gke_version
+
+  is_valid_master_version = (
+      master_gke_version in gke_server_config.valid_master_versions
+  )
+  is_valid_node_version = (
```
Comment (SurbhiJainUSC) [inline xpk.py:4173]: Why do we need to check the node pool gke version here? Isn't `get_gke_node_pool_version()` doing this?

Comment (Obliviour) [reply xpk.py:4188]: This is needed for cpu node pools created through cluster create command.

File: xpk.py (hunk 6 above)
Comment (Obliviour) [reply xpk.py]: fixed

File: xpk.py (hunk 5 above)
Comment (Obliviour) [reply xpk.py]: fixed

File: xpk.py (hunk 4 above)
Comment (Obliviour) [reply xpk.py]: Done! Thanks for the suggestion!

**Review by SurbhiJainUSC — APPROVED** [review]
